pkg/client/events/replay_client_integration_test.go

```diff
@@ -0,0 +1,147 @@
+	return t.Message
+}
+
+func newDecodeEventMessageFn() events.NewEventsFn[messageEvent] {
+	return func(eventBz []byte) (messageEvent, error) {
+		t := new(tEvent)
+		if err := json.Unmarshal(eventBz, t); err != nil {
+			return nil, err
+		}
+		if t.Message == "" {
+			return nil, events.ErrEventsUnmarshalEvent
+		}
+		return t, nil
+	}
+}
+
+// newMessageEventBz returns a new message event in JSON format
+func newMessageEventBz(eventNum int32) []byte {
+	return []byte(fmt.Sprintf(`{"message":"message_%d"}`, eventNum))
+}
+
+func TestReplayClient_Remapping(t *testing.T) {
+	var (
+		ctx               = context.Background()
+		connClosed        atomic.Bool
+		firstEventDelayed atomic.Bool
+		readEventCounter  atomic.Int32
+		eventsReceived    atomic.Int32
+		eventsToRecv      = int32(10)
+		errCh             = make(chan error, 1)
+		timeoutAfter      = 3 * time.Second // 1 second delay on retry.OnError
+	)
+
+	// Setup the mock connection and dialer
+	connMock, dialerMock := testeventsquery.NewNTimesReconnectMockConnAndDialer(t, 2, &connClosed, &firstEventDelayed)
+	// Mock the connection receiving events
+	connMock.EXPECT().Receive().
+		DoAndReturn(func() (any, error) {
+			// Simulate ErrConnClosed if connection is isClosed.
+			if connClosed.Load() {
+				return nil, events.ErrEventsConnClosed
+			}
+
+			// Delay the event if needed
+			if !firstEventDelayed.CompareAndSwap(false, true) {
+				time.Sleep(50 * time.Millisecond)
+			}
+
+			eventNum := readEventCounter.Add(1) - 1
+			event := newMessageEventBz(eventNum)
+			if eventNum == 2 {
+				// Simulate the connection closing
+				connMock.Close()
+			}
+
+			// Simulate IO delay between sequential events.
+			time.Sleep(50 * time.Microsecond)
+
+			return event, nil
+		}).
+		MinTimes(int(eventsToRecv))
+
+	// Setup the events query client dependency
+	dialerOpt := events.WithDialer(dialerMock)
+	queryClient := events.NewEventsQueryClient("", dialerOpt)
+	deps := depinject.Supply(queryClient)
+
+	// Create the replay client
+	replayClient, err := events.NewEventsReplayClient[messageEvent, messageEventReplayObs](
+		ctx,
+		deps,
+		"", // subscription query string
+		newDecodeEventMessageFn(),
+		100, // replay buffer size
+	)
+	require.NoError(t, err)
+
+	go func() {
+		// Subscribe to the replay clients events
+		eventTypeObs := replayClient.EventsSequence(ctx)
+		eventObserver := eventTypeObs.Subscribe(ctx)
+		var previousMessage messageEvent
+		for msgEvent := range eventObserver.Ch() {
+			var previousNum int32
+			var currentNum int32
+			if previousMessage != nil {
+				_, err := fmt.Sscanf(previousMessage.EventMessage(), "message_%d", &previousNum)
+				require.NoError(t, err)
+				_, err = fmt.Sscanf(msgEvent.EventMessage(), "message_%d", &currentNum)
+				require.NoError(t, err)
+			}
+			previousMessage = msgEvent
+
+			require.NotEmpty(t, msgEvent)
+			received := eventsReceived.Add(1)
+			if received >= eventsToRecv {
+				errCh <- nil
+				return
+			}
+		}
+	}()
```

Comment:
I think this can be expressed more and concisely as a `ForEach()`. Additionally, It looks like the event was being deserialized but for no real reason.

```suggestion
	channel.ForEach(
		ctx,
		observable.Observable[messageEvent](replayClient.EventsSequence(ctx)),
		func(ctx context.Context, msgEvent messageEvent) {
			require.NotEmpty(t, msgEvent)
			received := eventsReceived.Add(1)
			if received >= eventsToRecv {
				errCh <- nil
				return
			}
		},
	)
```